bootenv/keydata.go

```diff
@@ -0,0 +1,538 @@
+	alg := d.data.Params.ModelDigests.Alg
+	if !alg.Available() {
+		return
+	}
+
+	var modelDigests [][]byte
+	for _, model := range models {
+		digest, err := computeSnapModelHash(crypto.Hash(alg), model)
+		if err != nil {
+			return fmt.Errorf("cannot compute snap model digest: %w", err)
+		}
+		modelDigests = append(modelDigests, digest)
+	}
+
+	currentModelDigests := d.data.Params.ModelDigests.Digests
+	d.data.Params.ModelDigests.Digests = modelDigests
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		d.data.Params.ModelDigests.Digests = currentModelDigests
+	}()
+
+	return d.authorize(key, role)
+}
+
+// SetAuthorizedBootModes is used to set new authorized boot modes for existing key data scope.
+//
+// The PrimaryKey and the role parameters are used to derive a role unique signing key which is
+// used to sign a hash (using scope's MDAlg) of an ASN1 marshalled payload containing the already
+// authorized model digests and the new boot modes.
+// On error the scope's already authorized boot modes remain unchanged.
+func (d *KeyDataScope) SetAuthorizedBootModes(key secboot.PrimaryKey, role string, modes ...string) (err error) {
+	currentModes := d.data.Params.Modes
+	d.data.Params.Modes = modes
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		d.data.Params.Modes = currentModes
+	}()
+
+	return d.authorize(key, role)
+}
+
+// IsBootEnvironmentAuthorized checks if the current boot environment (model and boot mode) is
+// matches the key data's scope authorized models and boot modes.
+func (d *KeyDataScope) IsBootEnvironmentAuthorized() error {
+	ok, err := d.isAuthorized()
+	if err != nil {
+		return fmt.Errorf("cannot verify signature: %w", err)
+	}
+	if !ok {
+		return errors.New("invalid signature")
+	}
+
+	alg := d.data.Params.ModelDigests.Alg
+	if !alg.Available() {
+		return errors.New("model digest algorithm unavailable")
+	}
+
+	if len(d.data.Params.ModelDigests.Digests) > 0 {
+		model, err := loadCurrentModel()
+		if err != nil {
+			return err
+		}
+
+		currentModelDigest, err := computeSnapModelHash(crypto.Hash(alg), model)
+		if err != nil {
+			return fmt.Errorf("cannot compute snap model digest: %w", err)
+		}
+
+		found := false
+		for _, modelDigest := range d.data.Params.ModelDigests.Digests {
+			if bytes.Equal(modelDigest, currentModelDigest) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return errors.New("unauthorized model")
+		}
+	}
+
+	if len(d.data.Params.Modes) > 0 {
+		mode, err := loadCurrentBootMode()
+		if err != nil {
+			return err
+		}
+
+		found := false
+		for _, m := range d.data.Params.Modes {
+			if m == mode {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return errors.New("unauthorized boot mode")
+		}
+	}
+
+	return nil
+}
+
+// MakeAdditionalData constructs the additional data that need to be integrity protected for
+// a key data scope (in AES-GCM for example).
```

Comment:
this doc comment for this is very opaque, what should be done with the output of this? (the spec doesn't seem very clear on this either)

Reply:
I expanded it a little bit with text from the spec. My understanding is that this serves there as a reference implementation as nothing makes use of it for now. Pinging @chrisccoulson here too :)

Reply:
thanks, now I see that is additional data in the same e.g of https://pkg.go.dev/crypto/cipher@go1.22.0#AEAD ? I wonder though if the name should contain Cipher or AEAD: MakeAEADAdditionalData ?

Reply:
I think `MakeAEADAdditionalData` is probably ok, given that it's expected that another package uses an implementation of `cipher.AEAD` and consumes this.

Reply:
Renamed it as suggested, thanks.